mantidproject/mantid
# Pychop update

**Description of work.**

PyChop2 has been refactored to better integrate the two original codes for calculating the time widths (`CHOP` by T. G. Perring used to calculate widths for Fermi chopper instruments, and `MultiRep` by R. I. Bewley to calculate widths for LET). 

Previously these two codes (defined by functions in `chop.py` and `mulpyrep.py`) had individual wrapper classes, `ISISFermi` and `ISISDisk` respectively. However, the user was not supposed to call either of these classes directly, but was instead meant to use another class, `PyChop2`, which then called the methods of either `ISISFermi` or `ISISDisk` depending on which instrument and functionality was required (calculating rep-rate multiplication [RRM] reps always used `ISISDisk`, calculating the resolution and flux for MAPS, MARI and MERLIN used `ISISFermi` and for LET used `ISISDisk`). With MAPS and MARI now able to use RRM, this is not ideal, so the functionalities of `ISISFermi` and `ISISDisk` have been combined into a single new class `Instrument` which has the same API as `PyChop2` and replaces all three of the old classes. The old classes have been deprecated but still remain in the code.

Finally, instrument parameters (including chopper package information) used to be hard-coded into the classes as a `dict` keyed by the instrument name. This is now separated into a YAML file. `Instrument` can be initialised using either such a YAML file or a `dict` with the necessary information. YAML files for ISIS instrument (the information which was previously hard-coded) are now included in the PyChop folder and is automatically loaded by the GUI - so the behaviour should be the same as previously. The main reason for doing this is to allow extensions of PyChop to SNS instruments (ARCS, SEQUOIA and CNCS) to be a straightforward matter of adding a YAML for each instrument.

**To test:**



Run MantidPlot and check the interface runs. 

- Check that all the ISIS instruments (LET, MAPS, MARI, MERLIN) are in the `Instrument` combobox. Select one of them, and type a number in the `Ei` lineedit, and click `Calculate and Plot` - check that the output in the graph tabs (`Resolution`, `Flux-Ei`, `Flux-Freq`, etc.) look ok - The resolution should be around 1-5% of the incident energy, and should go down with increasing chopper frequency. 

- Click the `Hold current plot` checkbox and enter another `Ei` and click `Calculate and Plot` again. Check that the graph updates with the new energy (without overwriting the old one). 

- Uncheck `Hold current plot` and check `Show multi-reps` and click `Calculate and Plot`. You should see several different `Ei`s (reps), and these should correspond to the reps plotted in the `Time-distance` graph. 

- Set the instrument to one of MAPS / MARI / MERLIN. Set the energy to 1 meV and the chopper frequency to max (600Hz) and calculate. Check that a window pops up saying that there is zero transmission through the chopper. (For multi-reps, there should be several entries indicating which reps are not transmitted).

- Download: [sns_instruments.zip](https://github.com/mantidproject/mantid/files/2094871/sns_instruments.zip) and open one of the files, which contain **unverified** parameters for the SNS spectrometers. Load one of them in the GUI and check that all the calculations work like for the ISIS instruments. Reload the same file and check that a pop up appears asking if you want to overwrite the instrument in memory or load it as a different name.

- From a script window or IPython console run:

```
from PyChop import PyChop2, PyChopGui
merlin = PyChop2('merlin','s',250)
merlin.setEi(25)
resolution = merlin.getResolution()
print resolution
resolution, flux = PyChop2.calculate(inst='maps', package='a', freq=500, ei=600, etrans=range(0,550,50))
print resolution
print flux
```

and check that the output is:

```
[ 1.01445709  0.95917016  0.90562145  0.85388193  0.80403     0.75615228
  0.71034431  0.66671125  0.62536816  0.58643999  0.55006072  0.51637167
  0.48551853  0.45764732  0.43290012  0.41141295  0.39332259  0.37880081
  0.36818504]
[ 19.93042396  18.49629505  17.14386792  15.87801916  14.70404501
  13.62757936  12.65444134  11.79041481  11.04102026  10.41149001
   9.90757014]
[ 4407.97664496]
```

Fixes #22359. 

Does this update require release notes?


---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
